### PR TITLE
feat(zk): enable pkey generation

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements [keygen_pkxxx](https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/plonk/keygen.rs#L261-L407), while achieving both reducing code duplication and enhancing memory efficiency.

For reducing code duplication, it tries to share codes between `ProvingKey` and `VerifyingKey` by moving some shared parts to a parent class `Key`.

For enhancing memory efficiency, unlike Halo2, while computing `l_first`, `l_last` and `l_active_row`, it uses a single evaluations and recycles it.

### CODE REVIEW

**Comment by dongchangYoo** [comment]
I think it would be more natural for `tachyon/zk/plonk/circuit/assembly.h` to be located in `tachyon/zk/plonk/keys/assembly.h`

**Review by dongchangYoo — APPROVED** [review]
LGTM

File: tachyon/zk/plonk/keys/proving_key.h [hunk 1]
```diff
@@ -45,7 +39,127 @@ class ProvingKey {
...
+    return DoLoad(prover, std::move(pre_load_result), nullptr);
+  }
+
  private:
+  bool DoLoad(Prover<PCSTy>* prover, PreLoadResult&& pre_load_result,
+              VerifyingKeyLoadResult* vk_load_result) {
+    using Domain = typename PCSTy::Domain;
+
+    // NOTE(chokobole): |ComputeBlindingFactors()| is a second call. The first
+    // was called inside |PreLoad()|. But I think this is a cheap to compute.
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/keys/proving_key.h:71]:
```suggestion
    // was called inside |PreLoad()|. But I think this is cheap to compute.
```

**Review by TomTaehoonKim — APPROVED** [review]
Else, LGTM

**Pull request merged**